### DISCUSSION TITLE
`api` v7 — coming soon! 🤘🏽

### DISCUSSION THREAD

**Discussion by kanadgupta** [readmeio/api#791] [Polls]
Hey everyone! We wanted to provide y'all with an update on what we've been cooking up lately ⏲️

We've spent the last few weeks making lots of quality-of-life improvements to the code generation process and the generated SDKs themselves. We're still not quite finished yet, but we're excited to announce that the next major version of `api` is right around the corner!

Here's what you can expect:
- We've totally overhauled our SDK generation strategy so the generated SDK's types and module resolution should be consistent across all modern[^1] JavaScript environments (CommonJS, ES Modules, TypeScript, etc.). The CLI no longer asks confusing configuration questions about the type of JavaScript SDK you want to generate — it now generates a fully-typed JavaScript SDK that ✨ **_just works_** ✨ in pretty much every modern JavaScript environment.
- New `api` CLI commands to help you better manage your SDK installation(s) — commands to list, rebuild, uninstall your SDKs.
- Lots of little DX niceties (license generation, configuration files with sensible defaults, etc.) so you can generate and publish a high-quality SDK to [the NPM registry](https://www.npmjs.com/) within minutes 🚀 

Anything else you'd like to see? We'd love to hear your thoughts — feel free to leave a comment below! Also, if you could take a second to answer the poll below, we'd love to understand how you're using `api`-generated SDKs in practice.

Thanks for your feedback thus far — excited to see what y'all build with `api` v7!

-Kanad and the ReadMe team 🦉

[^1]: Note that `api` v7 will require Node.js v18 or greater if you're a Node.js user, and TypeScript 5.2 if you're a TypeScript user.

**Comment by jimmymcpeter** [comment]
I came across this library from [Wistia's docs](https://docs.wistia.com/reference/get_medias-media-hashed-id-captions-json) and was excited to try it.  I ran into issues using it from my ESM package in Node v18 and was ready to use raw node-fetch/axios until I saw your comment on #604 to use api@next.  I was able to use `npx api@next install https://dash.readme.com/api/v1/api-registry/1sia28loivyd0b` and now have a proper SDK with typings and autocomplete -- thank you for this upgrade

**Reply by kanadgupta** [reply]
Great to hear it's working for you @jimmymcpeter!

**Comment by elongberg** [comment]
Using this for the first time with [7shifts](https://developers.7shifts.com/reference/introduction), TypeScript 5.2 and commonjs module, had a few issues at first with the `api` 6.1.1 version but switched to `npx api@next install @7shifts/v2.2023.0501#1mj6ytblo5wg4q4` and everything just works!

Just a few changes I'd like to see so far:
1. The currently generated code exports a single instance of the SDK. However, we need multiple concurrent instances of the SDK each with a different `.auth(...)` setup. It's my understanding that the v6 api supported this with something like `const client = sdk(...)`, but that's no longer the case. We're working around this by tweaking the generated `index.ts`, but it would be nice if we didn't have to.
2. This is a bit of a special case, but `7Shifts` isn't a valid identifier. The generated code snippets to try out don't consider that, i.e.
```js
import 7Shifts from '@api/7shifts';

7Shifts.auth('token');
7Shifts.auth('token');
7Shifts.whoAmI()
  .then(({ data }) => console.log(data))
  .catch(err => console.error(err));
```
3. Perhaps related to the above: we've had to do some finagling to get the importing working. This is BEFORE any changes mentioned above:
```ts
import * as fauxAPI from "@api/7shifts";
const api = (fauxAPI as any).default as typeof fauxAPI;
api.auth(...);
```

**Comment by ben-pr-p** [comment]
Hi all, any timeline for release?

I'm getting this error: https://github.com/readmeio/api/issues/698

With `.postUsers_list` using `npx api install "@solidarity-tech/v1.0#qsba52hlzrqusc0" --lang ts --yes`, and not sure how to proceed.

**Comment by marcusrbrown** [comment]
Hello,

I need #820 to use `api`-generated packages in a Next.js project that uses the edge (WinterCG) runtime. Is this project still active?

**Reply by kanadgupta** [reply]
Hi there — [`v7.0.0-beta.11`](https://github.com/readmeio/api/releases/tag/v7.0.0-beta.11) includes this change. Appreciate your patience!

**Comment by joaocanaverde-blue** [comment]
@kanadgupta Been using v7 in a server context and it's amazing! I'd love to use this in Next.js from the browser, but it looks like I'll need #820 too. Let me know if I can help. 🙂

**Reply by kanadgupta** [reply]
Hi there — [`v7.0.0-beta.11`](https://github.com/readmeio/api/releases/tag/v7.0.0-beta.11) includes this change. Appreciate your patience!

**Comment by RopoMen** [comment]
I was generating SDK with the latest version and got following error while building. I also noticed that all generated files were `.ts` but imports were done explicitly as `.js` and nothing worked.
 
```
✖ Compiling your SDK
Command failed with exit code 1: npm run prepare
✘ [ERROR] Expected ";" but found "with"

    src/sdk.ts:4:41:
      4 │ import definition from '../openapi.json' with {
        │                                          ~~~~
        ╵                                          ;

✘ [ERROR] Expected ";" but found "with"

    src/sdk.ts:4:41:
      4 │ import definition from '../openapi.json' with {
        │                                          ~~~~
        ╵                                          ;

ESM Build failed
Error: Build failed with 1 error:
src/sdk.ts:4:41: ERROR: Expected ";" but found "with"
    at failureErrorWithLog (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:1649:15)
    at /Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:1058:25
    at runOnEndCallbacks (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:1484:45)
    at buildResponseToResult (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:1056:7)
    at /Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:1085:16
    at responseCallbacks.<computed> (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:703:9)
    at handleIncomingPacket (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:762:9)
    at Socket.readFromStdout (/Users/XYZ/repo123/node_modules/tsup/node_modules/esbuild/lib/main.js:679:7)
    at Socket.emit (node:events:519:28)
    at addChunk (node:internal/streams/readable:559:12)
CJS Build failed
```

**Reply by RopoMen** [reply]
Ok, issue comes from this line in sdk.ts `import definition from '../openapi.json' with { type: 'json' };` which is ok when Node.js 22 is used, but when used older Node versions (like mine 20) it wont work.

https://stackoverflow.com/questions/78759608/how-to-import-json-files-in-node-v22

So, solutions are
1. Support only node22+
2. Add flag to tell if legacy node is used.
3. Discover what is the used engine (if available)
4. Read JSON with fs module.

**Reply by RopoMen** [reply]
Hm..

Or is the issue that this is not one liner

```
import definition from '../openapi.json' with {
  type: 'json'
};

**Reply by kanadgupta** [reply]
Hi there — can you try reproducing with [`api@7.0.0-beta.13`](https://github.com/readmeio/api/releases/tag/v7.0.0-beta.13)? We've bumped the minimum required version of Node.js to v20.10.0[^1] and bumped several dependencies in that version (namely `tsup`) so hopefully import attributes should work as expected.

[^1]: Import attributes are supported in Node v20.10.0 and above. Compatibility info can be found on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import/with#browser_compatibility) and [the Node.js docs](https://nodejs.org/docs/latest-v22.x/api/esm.html#import-attributes).

**Reply by RopoMen** [reply]
Tested
```
npx api@v7.0.0-beta.13 install .api/myAPI.yaml --identifier myApi                                                           [2025/04/8T8:57:27]
Need to install the following packages:
api@7.0.0-beta.13
Ok to proceed? (y) y
✔ Fetching your API definition
✔ Generating your SDK
✔ Saving your SDK into your codebase
⚠ This generator requires some packages to be installed alongside it:
  › @readme/api-core: The core magic of your codegen'd SDK and is what is used for making requests. https://npm.im/api
  › tsup: Used for compiling your codegen'd SDK into code that can be used in JS environments. https://tsup.egoist.dev/
  › typescript: Required for `tsup`.
✔ OK to proceed with package installation? … yes
✖ Installing required packages
Command failed with exit code 1: npm install --save /Users/XYZ/REPOSITORY/packages/backend/.api/apis/aml
npm ERR! Cannot set properties of null (setting 'dev')
```

Generation goes through pretty ok, but still `index.ts` is importing `sdk.js` even though folder contains only `sdk.ts` and same goes to `schemas.ts` which imports `schemas/schemaxyz.js` and `schemas` folder contains only `.ts` files. So those imports cannot work.

**Reply by RopoMen** [reply]
I manually removed all `.js` extensions from imports but still there was issues in package.json level definition. I tested otherwise solution by importing `import myApi from 'packages/backend/.api/apis/myApi/src';` directly from generated stuff (with manual `.js` import fix) and then things started to work.

So generated code kind a works but there is still some issues to get it completely working.

**Reply by kanadgupta** [reply]
We're a little bit strapped for capacity to look into this at the moment, but a repository that reproduces this issue (that includes info on the Node.js and TypeScript versions you're using, as well as the OpenAPI definition you're attempting to build an SDK for) would be helpful if you could create one for us!

As for the `.js` file extensions that point to TypeScript files, that is by design and should work with any up-to-date Node/TypeScript setup. Matt Pocock has a good write-up on the reasoning here: https://www.totaltypescript.com/relative-import-paths-need-explicit-file-extensions-in-ecmascript-imports

**Reply by erunion** [reply]
I've definitely seen the `npm ERR! Cannot set properties of null (setting 'dev')` error before and if I recall right it's because we tried to run `npm` in a package with workspaces set up.

What version of `npm` do you have installed, or do you actually use `yarn` in your application?

**Reply by RopoMen** [reply]
We have npm workspaces in use as well. Unfortunately I cannot provide reference repository that would cause this issue. Simply I have no time for this.

**Comment by Magi1053** [comment]
Will this update include using correct package managers (yarn pnpm or bun) that is currently being used in the existing project?

**Reply by erunion** [reply]
It will, that is being worked on in https://github.com/readmeio/api/pull/1026

**Reply by Magi1053** [reply]
I have tried using beta 14 but I get the following error.

```

node:internal/modules/esm/resolve:873
  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), null);
        ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@readme/httpsnippet' imported from AppData\Local\npm-cache\_npx\3d1786b0703d9022\node_modules\httpsnippet-client-api\dist\index.js
    at packageResolve (node:internal/modules/esm/resolve:873:9)
    at moduleResolve (node:internal/modules/esm/resolve:946:18)
    at defaultResolve (node:internal/modules/esm/resolve:1188:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:642:12)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:591:25)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:574:38)
    at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:236:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:130:49) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v20.19.1
```

**Comment by bombillazo** [comment]
Hey there! I'm interested in this library and would like to contribute as much as possible. I want to know if it's being actively worked on or monitored so my time invested isn't time wasted (I've had this happen in other OS projects).

**Reply by erunion** [reply]
Hi! This library is unfortunately not being actively developed on right now aside from security fixes. The `next` channel is stable and production ready though!

**Reply by bombillazo** [reply]
Hmmm just curious, any reason? 

I really liked the simplicity, the quick integration with different API sources and how streamlined it worked, is this being replaced with some other tool?

**Reply by erunion** [reply]
It's not being replaced with some other tool right now, myself and Kanad have just been moved onto other projects internally for the time being.

**Reply by bombillazo** [reply]
Alright, big fan of the work you'all did :D

**Comment by erunion** [comment]
v7 is now live!
